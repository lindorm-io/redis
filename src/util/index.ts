export * from "./parse-blob";
export * from "./stringify-blob";
