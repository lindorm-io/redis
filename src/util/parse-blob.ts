import { TObject, isObject } from "@lindorm-io/core";
import { isArray, isBoolean, isObject as _isObject, isString } from "lodash";

const parseErrorValue = (input: any): Error => {
  const parsed = _isObject(input) ? input : JSON.parse(input);

  const error: any = new Error(parsed.message);
  error.stack = parsed.stack;

  for (const [key, value] of Object.entries(parsed)) {
    if (key === "message" || key === "stack") continue;
    error[key] = value;
  }

  return error;
};

const parseArrayValue = (input: any, meta: TObject<any>): Array<any> => {
  const parsed = isArray(input) ? input : JSON.parse(input);

  const result: Array<any> = [];

  for (const [index, value] of parsed.entries()) {
    if (isObject(meta[index])) {
      result.push(parseObjectValue(value, meta[index]));
    } else if (isArray(meta[index])) {
      result.push(parseArrayValue(value, meta[index]));
    } else if (meta[index] === "boolean") {
      result.push(isBoolean(value) ? value : JSON.parse(value));
    } else if (meta[index] === "date") {
      result.push(new Date(value));
    } else if (meta[index] === "error") {
      result.push(parseErrorValue(value));
    } else if (meta[index] === "number") {
      result.push(parseInt(value, 10));
    } else if (meta[index] === "string") {
      result.push(value);
    } else if (meta[index] === "null") {
      result.push(null);
    } else if (meta[index] === "undefined") {
      result.push(undefined);
    }
  }

  return result;
};

const parseObjectValue = (json: TObject<any>, meta: TObject<any>): TObject<any> => {
  const result: TObject<any> = {};

  for (const [key, value] of Object.entries(json)) {
    if (isObject(meta[key])) {
      result[key] = parseObjectValue(json[key], meta[key]);
    } else if (isArray(meta[key])) {
      result[key] = parseArrayValue(json[key], meta[key]);
    } else if (meta[key] === "boolean") {
      result[key] = JSON.parse(value);
    } else if (meta[key] === "date") {
      result[key] = new Date(value);
    } else if (meta[key] === "error") {
      result[key] = parseErrorValue(value);
    } else if (meta[key] === "number") {
      result[key] = parseInt(value, 10);
    } else if (meta[key] === "string") {
      result[key] = isString(value) ? value : JSON.parse(value);
    } else if (meta[key] === "null") {
      result[key] = null;
    } else if (meta[key] === "undefined") {
      result[key] = undefined;
    }
  }

  return result;
};

export const parseBlob = (string: string): TObject<any> => {
  const { json, meta } = JSON.parse(string);
  return parseObjectValue(json, meta);
};
