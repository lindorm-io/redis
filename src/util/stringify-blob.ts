import { isObject } from "@lindorm-io/core";
import { isArray, isBoolean, isDate, isError, isNull, isNumber, isString, isUndefined } from "lodash";

export const getType = (value: any): string => {
  if (isArray(value)) return "array";
  if (isBoolean(value)) return "boolean";
  if (isDate(value)) return "date";
  if (isError(value)) return "error";
  if (isNull(value)) return "null";
  if (isNumber(value)) return "number";
  if (isString(value)) return "string";
  if (isUndefined(value)) return "undefined";
};

export const getMetaArray = (input: Array<any>): Array<string> => {
  const result: Array<any> = [];

  for (const value of input) {
    if (isObject(value)) {
      result.push(getMetaObject(value));
    } else if (isArray(value)) {
      result.push(getMetaArray(value));
    } else {
      result.push(getType(value));
    }
  }

  return result;
};

export const getMetaObject = (input: Record<string, any>): Record<string, string> => {
  const result: Record<string, any> = {};

  for (const [key, value] of Object.entries(input)) {
    if (isObject(value)) {
      result[key] = getMetaObject(value);
    } else if (isArray(value)) {
      result[key] = getMetaArray(value);
    } else {
      result[key] = getType(value);
    }
  }

  return result;
};

const stringifyArrayValues = (input: Array<any>): Array<any> => {
  const result: Array<any> = [];

  for (const value of input) {
    if (isObject(value)) {
      result.push(stringifyObjectValues(value));
    } else if (isArray(value)) {
      result.push(stringifyArrayValues(value));
    } else if (isDate(value)) {
      result.push(value.toJSON());
    } else if (isError(value)) {
      result.push(JSON.stringify(value, Object.getOwnPropertyNames(value)));
    } else if (isString(value)) {
      result.push(value);
    } else if (isNull(value)) {
      result.push("null");
    } else if (isUndefined(value)) {
      result.push("undefined");
    } else {
      result.push(JSON.stringify(value));
    }
  }

  return result;
};

const stringifyObjectValues = (input: Record<string, any>): Record<string, any> => {
  const result: Record<string, any> = {};

  for (const [key, value] of Object.entries(input)) {
    if (isObject(value)) {
      result[key] = stringifyObjectValues(value);
    } else if (isArray(value)) {
      result[key] = stringifyArrayValues(value);
    } else if (isDate(value)) {
      result[key] = value.toJSON();
    } else if (isError(value)) {
      result[key] = JSON.stringify(value, Object.getOwnPropertyNames(value));
    } else if (isString(value)) {
      result[key] = value;
    } else if (isNull(value)) {
      result[key] = "null";
    } else if (isUndefined(value)) {
      result[key] = "undefined";
    } else {
      result[key] = JSON.stringify(value);
    }
  }

  return result;
};

export const stringifyBlob = (object: Record<string, any>): string => {
  return JSON.stringify({
    json: stringifyObjectValues(object),
    meta: getMetaObject(object),
  });
};
