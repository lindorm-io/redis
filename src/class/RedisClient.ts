import redis, { ClientOpts } from "redis";
import { IRedisClient, TPromise } from "../typing";
import { isNumber } from "lodash";
import { parseBlob, stringifyBlob } from "../util";
import { promisify } from "util";

export interface IAsyncClient {
  del: TPromise<number>;
  get: TPromise<string>;
  keys: TPromise<string>;
  quit: TPromise<string>;
  set: TPromise<string>;
  setEx: TPromise<string>;
}

export class RedisClient implements IRedisClient {
  private _client: IAsyncClient;
  private _clientOptions: ClientOpts;

  constructor(options: ClientOpts) {
    this._clientOptions = options;
  }

  public async connect(): Promise<void> {
    const client: redis.RedisClient = redis.createClient(this._clientOptions);

    this._client = {
      del: promisify(client.del).bind(client),
      get: promisify(client.get).bind(client),
      keys: promisify(client.keys).bind(client),
      quit: promisify(client.quit).bind(client),
      set: promisify(client.set).bind(client),
      setEx: promisify(client.setex).bind(client),
    };
  }

  public async quit(): Promise<string> {
    return this._client.quit();
  }

  public async set(key: string, value: Record<string, any>, expiresInSeconds?: number): Promise<string> {
    let result: string;

    if (isNumber(expiresInSeconds)) {
      result = await this._client.setEx(key, expiresInSeconds, stringifyBlob(value));
    } else {
      result = await this._client.set(key, stringifyBlob(value));
    }

    return result;
  }

  public async get(key: string): Promise<Record<string, any>> {
    return parseBlob(await this._client.get(key));
  }

  public async getAll(pattern: string): Promise<Array<Record<string, any>>> {
    const keys = await this._client.keys(pattern);
    const array: Array<Record<string, any>> = [];

    for (const key of keys) {
      array.push(await this.get(key));
    }

    return array;
  }

  public async del(key: string): Promise<number> {
    return this._client.del(key);
  }
}
