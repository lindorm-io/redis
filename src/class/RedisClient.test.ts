import { RedisClient } from "./RedisClient";
import { stringifyBlob } from "../util";

const mockGet = jest.fn((key: string, cb: any): void => {
  cb(null, stringifyBlob({ key, mock: "blob", data: 12345 }));
});
const mockKeys = jest.fn((pattern: string, cb: any): void => {
  cb(null, ["key1", "key2"]);
});
const mockSet = jest.fn((key: string, val: string, cb: any): void => {
  cb(null, "OK");
});
const mockSetex = jest.fn((key: string, val: string, num: number, cb: any): void => {
  cb(null, "OK");
});
const mockDel = jest.fn((key: string, cb: any): void => {
  cb(null, 1);
});

jest.mock("redis", () => ({
  createClient: () => ({
    get: mockGet,
    keys: mockKeys,
    set: mockSet,
    setex: mockSetex,
    del: mockDel,
  }),
}));

describe("RedisClient", () => {
  const client = new RedisClient({ port: 1 });

  test("should set", async () => {
    await expect(client.set("key", { mock: "mock" })).resolves.toBe("OK");
    expect(mockSet).toHaveBeenCalledWith(
      "key",
      '{"json":{"mock":"mock"},"meta":{"mock":"string"}}',
      expect.any(Function),
    );
  });

  test("should set with expiry", async () => {
    await expect(client.set("key", { mock: true }, 100)).resolves.toBe("OK");
    expect(mockSetex).toHaveBeenCalledWith(
      "key",
      100,
      '{"json":{"mock":"true"},"meta":{"mock":"boolean"}}',
      expect.any(Function),
    );
  });

  test("should get", async () => {
    await expect(client.get("key")).resolves.toMatchSnapshot();
    expect(mockGet).toHaveBeenCalledWith("key", expect.any(Function));
  });

  test("should get all", async () => {
    await expect(client.getAll("pattern*")).resolves.toMatchSnapshot();
  });

  test("should delete", async () => {
    await expect(client.del("key")).resolves.toBe(1);
    expect(mockDel).toHaveBeenCalledWith("key", expect.any(Function));
  });
});
