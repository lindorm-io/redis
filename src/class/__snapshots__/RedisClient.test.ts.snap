// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RedisClient should get 1`] = `
Object {
  "data": 12345,
  "key": "key",
  "mock": "blob",
}
`;

exports[`RedisClient should get all 1`] = `
Array [
  Object {
    "data": 12345,
    "key": "key1",
    "mock": "blob",
  },
  Object {
    "data": 12345,
    "key": "key2",
    "mock": "blob",
  },
]
`;
